src/lib/SaferNames/NameCore.hs

```diff
@@ -0,0 +1,294 @@
+  cont @UnsafeMakeDistinctS (UnsafeMakeBinder freshName) (UnsafeMakeName freshName)
+  where
+    freshName = freshRawName "v" scope
+
+freshRawName :: D.Tag -> S.Set RawName -> RawName
+freshRawName tag usedNames = D.Name D.GenName tag nextNum
+  where
+    nextNum = case S.lookupLT (D.Name D.GenName tag bigInt) usedNames of
+                Just (D.Name D.GenName tag' i)
+                  | tag' /= tag -> 0
+                  | i < bigInt  -> i + 1
+                  | otherwise   -> error "Ran out of numbers!"
+                _ -> 0
+    bigInt = (10::Int) ^ (9::Int)  -- TODO: consider a real sentinel value
+
+projectName :: NameSet (n:=>:l) -> Name s l -> Either (Name s n) (Name s (n:=>:l))
+projectName (UnsafeMakeNameSet scope) (UnsafeMakeName rawName)
+  | S.member rawName scope = Right $ UnsafeMakeName rawName
+  | otherwise              = Left  $ UnsafeMakeName rawName
+
+-- proves that the names in n are distinct
+class Distinct (n::S)
+
+instance Distinct VoidS
+instance Distinct UnsafeMakeDistinctS
+
+-- === injections ===
+
+-- When we inline an expression, we lift it into a larger (deeper) scope,
+-- containing more in-scope variables. For example, when we turn this:
+--   let foo = \x. x + z
+--   in \y. foo y
+-- into this:
+--   \y. (\x. x + z) y
+-- The expression `\x. x + z`, initially in the scope `[z]`, gets injected into
+-- the scope `[z, y]`. To prove that we can do the injection as an coercion,
+-- it's sufficient to know (1) that the new scope contains distinct names, and
+-- (2) it extends the old scope. These are the `Distinct l` and
+-- `(InjectableB b, b)` conditions below.
+
+injectNames :: InjectableE e => InjectableB b
+            => Distinct l => b n l -> e n -> e l
+injectNames _ x = unsafeCoerceE x
+
+-- `injectNames` is the only function we actualy use in practice. The rest of
+-- this section is just about proving that it's valid to use `injectNames`, by
+-- offering tools to implement `Injectable` instances.
+
+-- To prove that a data type is injectable by coercion, we have to show that
+-- it's injectable without using coercion. `injectionProof` is never actually
+-- called at run time, but it should still be implemented to prove that it's
+-- possible. This isn't watertight. For example, you could write
+-- `injectionProof = injectNames` and it'd still type check!
+class InjectableE (e::E) where
+  injectionProofE :: ObservablyFresh n l -> e n -> e l
+
+class InjectableB (b::B) where
+  injectionProofB :: ObservablyFresh n n' -> b n l
+                  -> (forall l'. ObservablyFresh l l' -> b n' l' -> a)
+                  -> a
+
+  -- Not sure if this belongs here, but it doesn't make sense in `HasNamesB`
+  -- either. Maybe we need a separate type class?
+  boundNames :: b n l -> NameSet (n:=>:l)
+
+-- an `ObservablyFresh n l` means that scope `l` is a superset of scope `n` that
+-- doesn't shadow any of the names exposed by `n`.
+data ObservablyFresh (n::S) (l::S) = UnsafeMakeObservablyFresh
```

Comment:
Right, it's a weaker condition that's handy for the recursive calls.

By "expose a name" I mean "present the name as the rightmost/innermost/lexically-available version of that name". What do I mean by "version"? Good question. You've caught me fudging things here and I'd like to find a way to be more precise about this.

I've said that a scope is just a list of names, but the `ObservablyFresh` predicate only makes sense if we consider names to have some kind of "identity" alongside their actual value. It's easy to describe if we imagine that all the variables happen to have different types. For example:

```
-- l is observably fresh wrt n
n = [x:Int,         y:Float]
l = [x:Int, y:Bool, y:Float, z:Int]

-- l not observably fresh wrt n. The "new" `y:Bool` shadows the old `y:Float`.
n = [x:Int, y:Float]
l = [x:Int, y:Float, y:Bool, z:Int]

```

This comes up if we imagine inlining `foo` in the following expression:

```
x:Int = 1
foo = (\y:Float. x + y)
y:Bool = True
z:Int = 2
foo
```

Before inlining, the `x + y` expression is in the scope `[x:Int, y:Float]`. After inlining, it's in the larger (deeper) scope `[x:Int, y:Bool, z:Int, y:Float]`. But it's ok, because even though the new scope adds an extra non-unique name, `y:Bool`, it's shadowed by the one we care about, the `y:Float`.

Does that make sense? I'd love a way to describe this precisely without resorting to unique identities or types that happen to be different.

Reply:
I think I get what you mean. I'm a bit worried about the encoding, though. It seems like whether something is observably fresh depends on a particular interpretation of `n` and `l`, but that interpretation isn't checked by the Haskell compiler. So I'd be a bit concerned that maybe we'd misinterpret the object as proving some property that doesn't actually hold.

I can't immediately think of a better way to encode this, though...

Maybe it would make sense at least for understanding purposes if we thought of a nameset as an ordered *mapping*, instead of an ordered list? Or at least the type-level metadata for such a mapping?

Reply:
Ok, fleshing out that idea a bit more. It sounds like `Name`s already "contain" some other static information `s`. And that presumably uniquely identifies what it means to be the "same"? Or at least it could, in principle, if we chose `s` correctly. So would it be correct to say:
- there's always exactly one mapping from names to their static information (for any `v`?)
- the `n :: S` phantom type parameter uniquely identifies this mapping, so that if you have a `NameMap v n` it must be the "right" one (carefully maintained using unsafe stuff). Essentially, we say that every NameMap is semantically a singleton object that gives us run-time access to the mapping described by `n`.
- an `ObservablyFresh n l` is a proof that, if you have a `NameMap v l`, and a `NameMap v n`, then the second is ALWAYS a sub-mapping of the first; equivalently, you can pretend your `NameMap v l` is the singleton instance of `NameMap v n` without changing the results of any lookup.
- a `NameSet n` is just a cheap representation of the keys of the `NameMap`, but it's not the fundamental object we are reasoning about. Maybe it's isomorphic to the singleton `NameMap UnitV n` or something that maps every name to the unit. Of course, it's got to follow the same guarantees above, but that's kind of meaningless because every payload is identical so ObservablyFresh doesn't mean anything important here.

I think this is true based on what you told me in person? But maybe it makes sense for this to be the explicit formalism we are operating under, as well.

(On the other hand, I'm not sure I know what it means for this interpretation of a name set to be "distinct". Maybe I'm still missing something, or maybe this interpretation isn't quite right either.)

Reply:
Yes, that sounds right to me. I'll add a longer comment.

Reply:
We're still missing a comment here